src/ema/math.rs

```diff
@@ -0,0 +1,114 @@
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_price_ema(iterations: u32, prev: Price, incoming: Price, smoothing: FixedU128) -> Price {
+    let (exp_smoothing, exp_complement) = exp_smoothing_and_complement(smoothing, iterations);
+    price_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the iterated exponential moving average for the given balances.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_balance_ema(iterations: u32, prev: Balance, incoming: Balance, smoothing: FixedU128) -> Balance {
+    let (exp_smoothing, exp_complement) = exp_smoothing_and_complement(smoothing, iterations);
+    balance_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the iterated exponential moving average for the given volumes.
+/// `iterations` is the number of iterations of the EMA to calculate.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `smoothing` is the smoothing factor of the EMA.
+pub fn iterated_volume_ema(
+    iterations: u32,
+    prev: (Balance, Balance, Balance, Balance),
+    incoming: (Balance, Balance, Balance, Balance),
+    smoothing: FixedU128,
+) -> (Balance, Balance, Balance, Balance) {
+    let (exp_smoothing, exp_complement) = exp_smoothing_and_complement(smoothing, iterations);
+    volume_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the smoothing factor and its complement for a given combination of period and
+/// iterations.
+pub fn exp_smoothing_and_complement(smoothing: FixedU128, iterations: u32) -> (FixedU128, FixedU128) {
+    debug_assert!(smoothing <= Price::one());
+    let complement = Price::one() - smoothing;
+    // in order to determine the iterated smoothing factor we exponentiate the complement
+    let exp_complement = complement.saturating_pow(iterations as usize);
+    debug_assert!(exp_complement <= Price::one());
+    let exp_smoothing = Price::one() - exp_complement;
+    (exp_smoothing, exp_complement)
+}
+
+/// Calculates smoothing factor alpha for an exponential moving average based on `period`:
+/// `alpha = 2 / (period + 1)`. It leads to the "center of mass" of the EMA corresponding to be the
+/// "center of mass" of a `period`-length SMA.
+///
+/// Possible alternatives for `alpha = 2 / (period + 1)`:
+/// + `alpha = 1 - 0.5^(1 / period)` for a half-life of `period` or
+/// + `alpha = 1 - 0.5^(2 / period)` to have the same median as a `period`-length SMA. See
+/// https://en.wikipedia.org/wiki/Moving_average#Relationship_between_SMA_and_EMA (N = period).
+pub fn smoothing_from_period(period: u64) -> FixedU128 {
+    FixedU128::saturating_from_rational(2u64, period.max(1).saturating_add(1))
```

Comment:
If we are storing & using the smoothing factor directly instead of the period (as it looks like we are), I care a lot less about this. Providing this utility function is fine.

Reply:
I'm averse to using the smoothing factor as a key in storage, but there's things we can do about that.

Reply:
What do you mean by "as a key in storage"?

Reply:
Sorry, not talking about this PR, but referencing https://github.com/galacticcouncil/warehouse/pull/76 which is connected.
There I use the period to identify the different oracles in storage. If we get rid of of the period in between the Oracle identifier (e.g. "Day") and the smoothing factor, then I would need another storage key. For reasons I have trouble articulating I don't think the smoothing factor is a great choice. In the end it's a technical detail I can figure out, so no need to worry too much about it.